Event list in session recordings player

alexkim205 opened the pull request:
## Changes

Lays the groundwork for Events card in #6487. Safe to merge as it's hidden behind its own FF. 

Design has changed since that issue was created—please refer to Figma here for most up-to-date mockups - https://www.figma.com/file/gQBj9YnNgD8YW4nBwCVLZf/PostHog-App?node-id=3753%3A28483

**Before**

<img width="1654" alt="Screen Shot 2021-11-09 at 3 32 23 PM" src="https://user-images.githubusercontent.com/13460330/141022414-002709a4-2665-4be9-9155-fb80a55b8878.png">

**After**

https://user-images.githubusercontent.com/13460330/141351015-7dc79884-4b64-485d-9f0f-011072919cd5.mov

This PR is part 1 of implementing the events list component in our new session recording playback experience. It's far from being done, but I don't want this to balloon into a bigger PR than it already is and would like to get this merged in so that I can build on top of it with smaller PR's. Changes here include:

- [x] Adds new eventListLogic that keeps track of local filters + cell measurement cache
- [x] Events list is virtualized (only render events that are visible and lazy load the rest) to optimize speed for very long lists (lots of events)
- [x] Remove fullscreen capability (previously triggered with hotkey "F") until we figure out what the fullscreen experience should look like. @clarkus
- [x] Remove `screenfull package`
- [x] Allow filtering events with fuzzy query matching via Fuse.js 
- [x] Add bolding logic to utils that resolve name to display from event (and corresponding tests)
- [x] Remove box shadows and white out GUI while keeping video background dark.
- [x] Various other stylistic changes to bring it up to date with new updates in Figma
- [x] Change the way eventsData is mutated and flows from parent logic to component logics.

Future PR's will:
- [ ] Implement auto scrolling to event being played
- [ ] Different event categories (error, highlight, active)
- [ ] More complex recording filters (exclusion, highlighed, etc.)
- [ ] Click event in list to jump to timestamp

Any feedback on code quality/design implementation is appreciated!

## How did you test this code?

- [x] Add logic tests for `eventsListLogic`
- [x] Add util tests

alexkim205 committed:
remove full screen and implement event list

alexkim205 committed:
Merge remote-tracking branch 'origin' into feat/player-events-recordings

alexkim205 committed:
implement event list

alexkim205 committed:
hide events list behind FF

alexkim205 committed:
Merge remote-tracking branch 'origin' into feat/player-events-recordings

rcmarron commented on frontend/src/scenes/session-recordings/sessionRecordingLogic.ts:
```diff
@@ -183,7 +198,34 @@ export const sessionRecordingLogic = kea<sessionRecordingLogicType>({
+                return filters?.query
+                    ? new Fuse(events, {
+                          threshold: 0.3,
+                          keys: ['event'],
```
I think we should allow searching on the displayed event text too, so you can search for "Continue" and see the autocapture event that says 'Clicked span with text "Continue"'

alexkim205 replied:
good catch! had this in the back of my mind but forgot to implement it

rcmarron commented on frontend/src/scenes/session-recordings/player/eventsListLogic.test.ts:
```diff
@@ -0,0 +1,41 @@
+                .toDispatchActions([sessionRecordingLogic.actionCreators.setFilters(filters)])
+                .toNotHaveDispatchedActions([sessionRecordingLogic.actionCreators.setFilters(filters)])
+        })
+        it('calls setFilter in parent logic without debounce', async () => {
```
I might be missing something here, but is there a difference between these two tests?

alexkim205 replied:
hmm looks like a brainfart

rcmarron commented on frontend/src/scenes/session-recordings/player/Seekbar.scss:
```diff
@@ -113,8 +113,8 @@ $tick_hover_size: 5px;
-                border: 0.5px solid $recording_player_container_bg;
-                background-color: $bg_light;
+                border: 0.5px solid $bg_light;
+                background-color: $recording_player_container_bg;
```
This looks kind of odd when events are clumped closely
![image](https://user-images.githubusercontent.com/4813045/141203882-de83d277-bd05-4797-8858-b39447ad641b.png)

alexkim205 replied:
Agreed, I think a darker border on lighter bg would make individual ticks stand out more. @clarkus What do you think?

rcmarron commented on frontend/src/scenes/session-recordings/player/PlayerEvents.tsx:
```diff
@@ -1,5 +1,72 @@
+            if (event.event.startsWith('$')) {
+                return <EventIcon />
+            }
+            return <ActionIcon />
```
I think this would give custom events the action icon. I think we were looking at using the 'code' icon for those.

alexkim205 replied:
There doesn't seem to be a way to distinguish events, custom_events, and actions from the events results that are returned from the `api/events` endpoint. For now, I'll default all non pageevents or autocapture events to show the event icon, and make a note to cut a PR that attaches the `action_id` to all events returned from this endpoint.

rcmarron commented on frontend/src/scenes/session-recordings/SessionPlayerDrawer.tsx:
```diff
@@ -27,7 +27,11 @@ export function SessionPlayerDrawer({ isPersonPage = false, onClose }: SessionPl
                 <Col style={{ height: '100vh' }}>
-                    <Row style={{ height: 40 }} align="middle" justify="space-between">
+                    <Row
+                        style={{ height: 40, borderBottom: '1px solid var(--border)' }}
```
Cool! Had no idea you could use css variables inline like that!

rcmarron commented on frontend/src/lib/utils.tsx:
```diff
@@ -619,19 +619,45 @@ export function isEmail(string: string): boolean {
 }
 
-export function eventToName(event: Pick<EventType, 'elements' | 'event' | 'properties'>): string {
+export function eventToName(
```
Nit: maybe `eventToDescription` is more appropriate? Feels like the event name is `$pageview` etc. Same for `autoCaptureEventToName`

rcmarron reviewed (dismissed):
Looking great! A couple of thoughts:

* I think we should move to the slightly different design where the event's name is still displayed in the list.
* On my screen, it feels like we're giving too much horizontal space to the list. Maybe we put a max width on it. 
![image](https://user-images.githubusercontent.com/4813045/141205503-eecc59a7-c86c-4194-9ddd-0e425a7fe834.png)
* We might not be sorting the events by time
![image](https://user-images.githubusercontent.com/4813045/141204194-0d2e4503-80d3-4cfa-a768-3652e51540af.png)

I'm excited to get the auto-scroll and 'click to seek' in there!

alexkim205 committed:
Merge remote-tracking branch 'origin' into feat/player-events-recordings

clarkus reviewed (dismissed):
The event list is looking good. We can set a max-width of 320px or so. We should also right-align the timestamp values. Users are going to compare from the right to the left with time values, so right aligning will help facilitate that.

alexkim205 commented:
> Looking great! A couple of thoughts:
> 
> * I think we should move to the slightly different design where the event's name is still displayed in the list.

👍 

> * On my screen, it feels like we're giving too much horizontal space to the list. Maybe we put a max width on it.

👍  put a 320 max-width on it as Chris recommended

> * We might not be sorting the events by time

It might be worth looking into why CH is returning unordered events since I don't think we're doing any special transformation in the frontend. 

> I'm excited to get the auto-scroll and 'click to seek' in there!

Me too!

alexkim205 committed:
events list card styles + fuse query options + more fixes

alexkim205 committed:
Merge remote-tracking branch 'origin' into feat/player-events-recordings

alexkim205 committed:
Merge remote-tracking branch 'origin' into feat/player-events-recordings

alexkim205 committed:
revert event item titles and fix responsiveness on mobile

alexkim205 commented:
@rcmarron addressed feedback. Also fixed mobile responsiveness bug where player wasn't expanding to full height:

https://user-images.githubusercontent.com/13460330/141351374-42bc8eb5-cc4d-4bfe-8bc8-47cc0dac3ba4.mov

rcmarron commented on frontend/src/scenes/session-recordings/sessionRecordingLogic.ts:
```diff
@@ -30,6 +31,22 @@ export const parseMetadataResponse = (metadata: Record<string, any>): Partial<Se
+const makeEventsQueryable = (events: SeekbarEventType[]): SeekbarEventType[] => {
+    return events.map((e) => {
+        let queryValue = e?.event ?? ''
+        if (['$pageview', '$pageleave'].includes(e.event)) {
```
Should we pull in `eventToDescription` here and use that so we're definitely filtering by what's displayed? Maybe we do something similar for the event name too?

alexkim205 replied:
good idea! especially now that we don't have to include bold JSX in `eventToDescription`

rcmarron reviewed (dismissed):
Looking super close! One comment and a couple of smaller things that i noticed:
1) For the description, can we extend it so it goes beneath the timestamp. Also, I think we may want to truncate instead of wrapping (I don't have a strong preference here, but I think it truncates in figma)
![image](https://user-images.githubusercontent.com/4813045/141364426-90a47e62-015b-474f-b96e-c277c2653faa.png)
2) Also on the description, I think the bold text is fighting with the event's name. Looks like Figma doesn't have the bolding
3) When we use the name as the description, it can feel pretty repetitive. I wonder if we should either have no description in that case or use the URL? @clarkus Do you have thoughts here?

![image](https://user-images.githubusercontent.com/4813045/141364844-af96558b-8d94-46ef-9a5d-4e8a83fd3f98.png)

clarkus commented:
> 1. For the description, can we extend it so it goes beneath the timestamp. Also, I think we may want to truncate instead of wrapping (I don't have a strong preference here, but I think it truncates in figma)

How about we wrap up to two rows and then truncate. I don't think we'll have descriptions longer than 2 rows and that's mostly going to depend on how verbose our autocapture content is. 

> 2. Also on the description, I think the bold text is fighting with the event's name. Looks like Figma doesn't have the bolding

👍 

> 3. When we use the name as the description, it can feel pretty repetitive. I wonder if we should either have no description in that case or use the URL? @clarkus Do you have thoughts here?

Is autocapture the only case where we need to show a description? This might be good justification to going back to the smaller-form version?

alexkim205 commented:
> 1. For the description, can we extend it so it goes beneath the timestamp. Also, I think we may want to truncate instead of wrapping (I don't have a strong preference here, but I think it truncates in figma)

Extending it beneath the timestamp makes sense. I would prefer wrapping over truncating since valuable information that can be used to differentiate events might be truncated. For example two autocaptures clicking on buttons with different text.

> 2. Also on the description, I think the bold text is fighting with the event's name. Looks like Figma doesn't have the bolding

👍

alexkim205 committed:
feedback

alexkim205 committed:
Merge remote-tracking branch 'origin' into feat/player-events-recordings

rcmarron reviewed (dismissed):
This is looking really really good. I noticed one last thing on the filtering. I can't figure it out exactly, but sometimes it seems to not include events that I'd expect to see. 

https://user-images.githubusercontent.com/4813045/141399885-a0f633be-3090-4994-bb29-3fc823bc7350.mov

alexkim205 committed:
remove quotes to fix tokenization

alexkim205 commented:
> This is looking really really good. I noticed one last thing on the filtering. I can't figure it out exactly, but sometimes it seems to not include events that I'd expect to see.

Huh it looks like Fuse doesn't tokenize strings that are between quotes, which is why it was impossible to match on the elements texts. https://github.com/krisk/Fuse/issues/235

Should be fixed now!

alexkim205 committed:
Merge remote-tracking branch 'origin' into feat/player-events-recordings

rcmarron committed:
small tweaks

rcmarron reviewed (approved):
Looks good!

I just made 2 very small changes. One formatting change on the list, and I moved the fuse threshold back to it's default. The text matching is a bit funny, and I think it's better to be less strict instead of more strict.
